src/spotify_to_tidal/sync.py

```diff
@@ -172,133 +172,159 @@ def get_tracks_from_spotify_playlist(spotify_session: spotipy.Spotify, spotify_p
-                results.append( (track, None) )
-        return (results, cache_hits)
+                track_match_cache.insert((spotify_track['id'], tidal_track.id))
+                spotify_tracks.pop(idx)
+                return
+
+    # make a copy of the tracks to avoid modifying original arrays
+    spotify_tracks = [t for t in spotify_tracks]
+    tidal_tracks = [t for t in tidal_tracks]
+
+    # first populate from the tidal tracks
+    for track in tidal_tracks:
+        _populate_one_track_from_tidal(track)
+    # then populate from the subset of Spotify tracks that didn't match (to account for many-to-one style mappings)
+    for track in spotify_tracks:
+        _populate_one_track_from_spotify(track)
+
+def get_new_tracks_from_spotify_playlist(spotify_tracks, old_tidal_tracks):
+    ''' Extracts only the new tracks in the Spotify playlist that are not already on Tidal or known match failures '''
+    populate_track_match_cache(spotify_tracks, old_tidal_tracks)
+    results = []
+    for spotify_track in spotify_tracks:
+        if not spotify_track['id']: continue
+        cached_track_id = track_match_cache.get(spotify_track['id'])
+        if not track_match_cache.get(spotify_track['id']) and not failure_cache.has_match_failure(spotify_track['id']):
+            results.append(spotify_track)
+    return results
 
-def tidal_playlist_is_dirty(playlist: tidalapi.Playlist, new_track_ids: Sequence[str]) -> bool:
-    old_tracks = playlist.tracks()
-    if len(old_tracks) != len(new_track_ids):
-        return True
-    for i in range(len(old_tracks)):
-        if old_tracks[i].id != new_track_ids[i]:
-            return True
-    return False
+def get_tracks_for_new_tidal_playlist(spotify_tracks):
+    ''' gets list of corresponding tidal track ids for each spotify track, ignoring duplicates '''
+    output = []
+    seen_tracks = set()
+    for spotify_track in spotify_tracks:
+        tidal_id = track_match_cache.get(spotify_track['id'])
+        if tidal_id and not tidal_id in seen_tracks:
+            output.append(tidal_id)
+            seen_tracks.add(tidal_id)
+    return output
 
-def sync_playlist(spotify_session: spotipy.Spotify, tidal_session: tidalapi.Session, spotify_id: str, tidal_id: int, config):
-    try:
-        spotify_playlist = spotify_session.playlist(spotify_id)
-    except spotipy.SpotifyException as e:
-        print("Error getting Spotify playlist " + spotify_id)
-        print(e)
-        return
-    if tidal_id:
-        # if a Tidal playlist was specified then look it up
-        try:
-            tidal_playlist = tidal_session.playlist(tidal_id)
-        except Exception as e:
-            print("Error getting Tidal playlist " + tidal_id)
-            print(e)
-            return
-    else:
-        # create a new Tidal playlist if required
+async def sync_playlist(spotify_session: spotipy.Spotify, tidal_session: tidalapi.Session, spotify_playlist, tidal_playlist: tidalapi.Playlist | None, config):
```

Comment:
Quite a bit of refactoring in this function now. The logic is basically the same, but now it takes the playlist objects directly instead of taking the playlist ids, in order to avoid redundant API calls. I also made it a coroutine instead of employing multiprocessing, and introduced a simple rate limiter with `asyncio.Semaphore` and leaky bucket algorithm.